polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1211,31 +1258,98 @@ impl<T: Config> Pallet<T> {
 		ensure!(assets.len() <= MAX_ASSETS_FOR_TRANSFER, Error::<T>::TooManyAssets);
 		let value = (origin_location, assets.into_inner());
 		ensure!(T::XcmReserveTransferFilter::contains(&value), Error::<T>::Filtered);
-		let (origin_location, assets) = value;
-		let context = T::UniversalLocation::get();
-		let fees = assets
-			.get(fee_asset_item as usize)
-			.ok_or(Error::<T>::Empty)?
-			.clone()
-			.reanchored(&dest, context)
-			.map_err(|_| Error::<T>::CannotReanchor)?;
-		let max_assets = assets.len() as u32;
-		let assets: MultiAssets = assets.into();
-		let xcm = Xcm(vec![
-			BuyExecution { fees, weight_limit },
-			DepositAsset { assets: Wild(AllCounted(max_assets)), beneficiary },
-		]);
-		let mut message = Xcm(vec![
-			SetFeesMode { jit_withdraw: true },
-			TransferReserveAsset { assets, dest, xcm },
-		]);
-		let weight =
-			T::Weigher::weight(&mut message).map_err(|()| Error::<T>::UnweighableMessage)?;
-		let hash = message.using_encoded(sp_io::hashing::blake2_256);
-		let outcome =
-			T::XcmExecutor::execute_xcm_in_credit(origin_location, message, hash, weight, weight);
-		Self::deposit_event(Event::Attempted { outcome });
-		Ok(())
+		let (origin_location, mut assets) = value;
+
+		if fee_asset_item as usize >= assets.len() {
+			return Err(Error::<T>::Empty.into())
+		}
+		let mut fees = assets.swap_remove(fee_asset_item as usize);
+		let fees_transfer_type =
+			<T::XcmExecutor as AssetTransferSupport>::determine_for(&fees, &dest)
+				.map_err(Error::<T>::InvalidAsset)?;
+		let assets_transfer_type = if assets.is_empty() {
+			// Single asset to transfer (one used for fees where transfer type is determined above).
+			ensure!(fees_transfer_type != TransferType::Teleport, Error::<T>::Filtered);
+			fees_transfer_type
+		} else {
+			// Find reserve for non-fee assets.
+			Self::validate_assets_and_find_reserve(&assets, &dest)?
+		};
+
+		// Disallow (for now) different _remote_ reserves for assets and fees.
+		match (&fees_transfer_type, &assets_transfer_type) {
+			(TransferType::RemoteReserve(a), TransferType::RemoteReserve(b)) if a != b =>
+				return Err(Error::<T>::TooManyReserves.into()),
+			_ => (),
+		};
+
+		if fees_transfer_type == assets_transfer_type {
+			// Same reserve location (fees not teleportable), we can batch together fees and assets
+			// in same reserve-based-transfer.
+			assets.push(fees.clone());
+		} else {
+			// Different transfer types: we have to do separate transfers for fees and assets.
+			// This code block handles transferring the assets *used for fees*.
+
+			// Use only half the weight limit for fees transfer, then other half for assets
+			// transfer.
+			weight_limit = Self::halve_weight_limit(&weight_limit);
```

Comment:
I'm not sure if we should split "max_weight" too, or if splitting the "fees" amount is enough..

Looking for reviewers' opinion.